src/com.docker.slirp/src/main.ml

```diff
@@ -120,86 +184,131 @@ let main_t socket_path port_control_path max_connections vsock_path db_path nofi
-  Host.Main.run @@ main_t socket port_control max_connections vsock_path db nofile pcap debug
+      None in
+
+  let uri = Uri.of_string socket_url in
+  match Uri.scheme uri with
+  | Some "hyperv-connect" ->
+    let module Slirp_stack = Slirp.Make(Config)(Vmnet.Make(HV))(Resolv_conf)(Host) in
+    let sockaddr = hvsock_addr_of_uri ~default_serviceid:ethernet_serviceid (Uri.of_string socket_url) in
+    hvsock_connect_forever socket_url sockaddr
+      (fun fd ->
+        let conn = HV.connect fd in
+        ( match config with
+          | Some config -> Slirp_stack.create config
+          | None -> Lwt.return hardcoded_configuration )
+        >>= fun stack ->
+        Slirp_stack.connect stack conn
+        >>= fun stack ->
+        Log.info (fun f -> f "stack connected");
+        Slirp_stack.after_disconnect stack
+        >>= fun () ->
+        Log.info (fun f -> f "stack disconnected");
+        Lwt.return ()
+      )
+  | _ ->
+    let module Slirp_stack = Slirp.Make(Config)(Vmnet.Make(Host.Sockets.Stream.Unix))(Resolv_conf)(Host) in
+    unix_listen socket_url
+    >>= fun server ->
+    Host.Sockets.Stream.Unix.listen server
+      (fun conn ->
+        ( match config with
+          | Some config -> Slirp_stack.create config
+          | None -> Lwt.return hardcoded_configuration )
+        >>= fun stack ->
+        Slirp_stack.connect stack conn
+        >>= fun stack ->
+        Log.info (fun f -> f "stack connected");
+        Slirp_stack.after_disconnect stack
+        >>= fun () ->
+        Log.info (fun f -> f "stack disconnected");
+        Lwt.return ()
+      );
+    let wait_forever, _ = Lwt.task () in
+    wait_forever
 
+let main socket_url port_control_url max_connections vsock_path db_path dns pcap debug =
+  Host.Main.run
+    (main_t socket_url port_control_url max_connections vsock_path db_path dns pcap debug)
 end
 
-let main socket port_control max_connections vsock_path db nofile pcap select debug =
+let main socket port_control max_connections vsock_path db_path dns pcap select debug =
   let module Use_lwt_unix = Main(Host_lwt_unix) in
   let module Use_uwt = Main(Host_uwt) in
   (if select then Use_lwt_unix.main else Use_uwt.main)
-    socket port_control max_connections vsock_path db nofile pcap debug
+    socket port_control max_connections vsock_path db_path dns pcap debug
 
 open Cmdliner
 
-(* NOTE(aduermael): it seems to me that "/var/tmp/com.docker.slirp.socket" is a default value, right?
-This socket path is now dynamic, depending on current user's home directory. Could we just
-make it fail instead? In case no argument is supplied? *)
 let socket =
-  Arg.(value & opt string "/var/tmp/com.docker.slirp.socket" & info [ "socket" ] ~docv:"SOCKET")
+  let doc =
+    Arg.info ~doc:
+      "A URLs to connect to for ethernet of the form \
+      hyperv-connect://vmid/serviceid or hyperv-connect://vmid for the default serviceid on Windows \
+      or /var/tmp/com.docker.slirp.socket on Mac" ["ethernet"]
```

Comment:
I don't think what I've written makes any sense. I think I composed together words found in both the Mac and Windows forks, without checking the result for grammatical correctness.